### PR TITLE
Add non-sequential request ID to Protocol errors

### PR DESCRIPTION
**Pull request by SunFulong**
As discussed in #436 , adding message with non-sequential request ID to [protocol errors](https://wamp-proto.org/wamp_latest_ietf.html#name-protocol-errors) scenarios.

### CODE REVIEW

**Review by oberstet — APPROVED** [review]

**Comment by oberstet** [comment]
The added text clarifies behavior, I'm +1 on merging this. Even more explicit: the new text line could reference the section "Session Scope IDs" in this same doc. But I'm fine with the change as is as well. @ecorm @KSDaemon what do you think in light of our recent discussions?

**Comment by SunFulong** [comment]
> Even more explicit: the new text line could reference the section "Session Scope IDs" in this same doc.

Sorry, I don't quite get what you mean. Should I add a reference to [that section](https://wamp-proto.org/wamp_latest_ietf.html#section-2.1.2-11) or to [sequential definition](https://wamp-proto.org/wamp_latest_ietf.html#section-2.1.2-4.3)?

BTW, I don't think messages from peer responses should be listed in the "Session Scope ID" section, their request IDs are copied from the original request, not newly generated.

**Comment by oberstet** [comment]
> Should I add a reference to [that section](https://wamp-proto.org/wamp_latest_ietf.html#section-2.1.2-11) or to [sequential definition](https://wamp-proto.org/wamp_latest_ietf.html#section-2.1.2-4.3)?

yeah, that's what I meant (sorry for confusion)

> their request IDs are copied from the original request, not newly generated.

yes, that's correct: we could add a note in the section that clarifies it - I wouldn't remove the messages from the list though, they do contain session scoped IDs, just that those IDs were generated by the other peer in the first place ..

File: rfc/text/basic/bp_02_building_blocks.md [hunk 1]
```diff
@@ -279,6 +279,7 @@ Following scenarios have to be considered protocol errors:
  - Receiving `REGISTERED` message, before session was established.
  - Receiving `UNREGISTERED` message, before session was established.
  - Receiving `INVOCATION` message, before session was established.
+ - Receiving message with non-sequential request ID, such as `SUBSCRIBE`, `UNSUBSCRIBE`, `PUBLISH`, `REGISTER`, `UNREGISTER`, `CALL` and `YIELD`.
```
Comment (KSDaemon) [inline rfc/text/basic/bp_02_building_blocks.md:282]: Well, here are listed only messages from client peers to the router. But I think this also applies to the reverse direction.

Comment (KSDaemon) [reply rfc/text/basic/bp_02_building_blocks.md:282]: Hm... Thinking... Are there any messages that can be initiated by router?... Seems that no. Only responses. So, sorry for the confusion.

**Review by KSDaemon — APPROVED** [review]

**Comment by KSDaemon** [comment]
Let's add a comment about copying the `request id`. I think that would be useful. And then merge!

**Comment by ecorm** [comment]
Looks fine to me. Under the subsection title, shouldn't it be:

> WAMP implementations MUST ~close~ abort sessions

"close" to me implies cleanly shutting down the session with a GOODBYE exchange, whereas "abort" implies sending an `ABORT` message with the protocol error URI and then disconnecting.

**Comment by ecorm** [comment]
> Let's add a comment about copying the request id

Did you mean duplicate request IDs?

**Comment by KSDaemon** [comment]
> Did you mean duplicate request IDs?

I mean the copying `Request|id` from `SUBSCRIBE` → `SUBSCRIBED` and so on

**Comment by oberstet** [comment]
"duplicate" or "copy" request IDs both might suggest there is yet another, different ID involved .. which of course it isn't. a 3rd option could be: "request ID mirrored in reply". well, that's a mouthful, not 1 verb of course;)

**Comment by KSDaemon** [comment]
Ou yeah! I was also thinking of `mirroring` :)

**Comment by SunFulong** [comment]
Sorry for the rubbish log, I don't know only one reviewer can be requested to re-review before.

**Comment by oberstet** [comment]
Don't worry about those "request review GH itches" .. we will gather feedback and reviews, we want this to be merged - once we agree on the exact details;) IOW: I'm kindly asking you for some patience, your input is much appreciated and we want to take the chance to really "fix it".

I've looked into the whole spec sections resulting again, and I just now (sorry!) one more problem:

1. @ecorm correctly noted that the wording should be "abort" rather than "close" when encountering protocol errors (this is described in the paragraph "In all such cases WAMP implementations:") - the PR reflects that now
2. BUT: the [description of ABORT](https://wamp-proto.org/wamp_bp_latest_ietf.html#name-abort-2) only has a message sequence chart that shows ABORT as being used as part of the opening of a WAMP session, and the whole ABORT message is only introduced as a subsection of "4.1 Session Establishment"

> also note that WELCOME correctly is described with its use as "completes the opening of a WAMP session", while ABORT (in the same "Session Establishment" section) is described with its use as "may abort a WAMP session". IOW: only the opening, or opening and later on as well?

If ABORT is a general alternative to GOODBYE as described in "4.2 Session Closing", and not only as part of opening a WAMP session, this should be made clear, at least by having a respective message sequence diagram for that as well.

-----

another problem is the missing explanation of why.

I mean (this is my view, lets discuss!): 

* yes, we need an ABORT message different from WELCOME in the opening of a session, as we want to be able to orderly deny a client that didn't do anything bad at the protocol level
* yes, we want a peer to "hard drop" a session by default whenever the peer did violate the protocol at a basic level, because a) continue talking to a bad (=buggy) client opens additional security risks, and b) because of [Maintaining Robust Protocols](https://datatracker.ietf.org/doc/html/draft-iab-protocol-maintenance): the Internet new-school

and we want to do the latter even though there is:

"be conservative in what you do, be liberal in what you accept from others": [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle) - the Internet old-school

-----

"hard drop": https://wamp-proto.org/wamp_bp_latest_ietf.html#name-protocol-errors describes this in

> SHOULD also drop the WAMP connection at transport level (recommended to prevent denial of service attacks)

what isn't mentioned is: _if_ and implementation chooses to follow the SHOULD recommendation and "hard drop" a connection, then no ABORT message is sent at all!

**Comment by SunFulong** [comment]
> I'm kindly asking you for some patience

I have chosen to use the WAMP protocol as the core of my software architecture for the next few years, so I am very concerned about this specification.

But I have a lot of work on weekdays, so the progress may be a bit slow. I usually free on weekends, please wait for me a few more days, I will definitely be back.

**Review by KSDaemon — APPROVED** [review]
For me this already looks good and pretty clear 👍🏻

**Comment by oberstet** [comment]
@KSDaemon from my PoV, there are only 2 details lacking precision in the spec (this of course isn't a mistake in this PR, but sth we could fix within this PR as well since we are at it):

**1)**
> If ABORT is a general alternative to GOODBYE as described in "4.2 Session Closing", and not only as part of opening a WAMP session, this should be made clear, at least by having a respective message sequence diagram for that as well.

**2)**
> what isn't mentioned is: if an implementation chooses to follow the SHOULD recommendation and "hard drop" a connection, then no ABORT message is sent at all! Or is an implementation still required to follow "MUST send an ABORT message to the offending peer, having reason wamp.error.protocol_violation " ?

What's your take / interpretation?

**Comment by KSDaemon** [comment]
Oh, sorry, missed it.
1) Agree! We should update the spec with appreciative diagram changes
2) Well... What is the current behavior if a peer receives an incorrect message type (like getting a second WELCOME or EVENT before the session was established)? It `MUST` send an ABORT message with `wamp.error.protocol_violation` reason and close connection. Right? Receiving non-incremental Request ID should work the same. I think it is just another type of `protocol_violation`. So I'm +1 on `MUST send an ABORT message`.

**Comment by ecorm** [comment]
One thing that isn't clear to me: when using an `ABORT` during the `HELLO` handshake, should the underlying transport be disconnected afterwards, or left connected? The same question arises when using `ABORT` for protocol errors.

 > Or is an implementation still required to follow "MUST send an ABORT message to the offending peer, having reason wamp.error.protocol_violation " ?

I think we must ask ourselves what are the pros and cons of sending an ABORT in the case of protocol errors, versus terminating the transport connection.

Pros for sending abort
- Helps client developers in troubleshooting the reason why a client connection was dropped. However the same information could also be stored in router-side logs.
- Disambiguates protocol error situations versus the transport being disconnected due to network failures.

Cons for sending abort
- If transport disconnection is mandated after sending an ABORT, it can be tricky to implement in a manner that doesn't cut off the ABORT message before it is fully received by the client.
- Additional workload for the router in a DoS attack.

**Comment by KSDaemon** [comment]
Good points @ecorm! 
And I still think it will be useful to send an `ABORT` message to the router. I suppose that in most high-level languages/frameworks if you issue a kind of `send` function call your app will get the control back after the message will be delivered somewhere into the OS/cache/network/buffers. So even if you close the network connection from your high-level app code, the message will already be in network subsystem queues or even on the hardware level which can handle retransmission or some other possible problems. Anyway, I don't think we should really go that deep. 
In the worst case, the abort message will just not be sent at all so the router will just close and clear the connection on its side.

**Comment by oberstet** [comment]
> I think we must ask ourselves what are the pros and cons of sending an ABORT in the case of protocol errors, versus terminating the transport connection.

yes, exactly!

> Pros for sending abort

yep, that's a good list, and fwiw, I'm +1 on this option (and that would make AGREE into a message _only_ used in the opening handshake, and hence no additional sequence message chart would be needed as well)

>  if you issue a kind of send function call your app will get the control back after the message will be delivered somewhere into the OS/cache/network/buffers. 

I would argue: development is commonly done using loopback or LAN. in such a situation, when the router drops the TCP, the app will notice immediately (not via a failed WAMP call or what, but via a failed/dropped TCP transport).

If the app development is using a router over WAN, then the app developer _should_ setup client-initiated heartbeating anyways, and if so, a dropped connection will be detected in the app (on the next heartbeat send)

Only if the app development is over WAN and not using client-initiated heartbeating, the router sending ABORT could speed up the app detecting what happened.

**Comment by ecorm** [comment]
It seems issue #255 isn't really resolved if we're considering the use of ABORT messages in the event of protocol errors. In that issue, @oberstet had advocated abrupt disconnection.

> I suppose that in most high-level languages/frameworks if you issue a kind of send function call your app will get the control back after the message will be delivered somewhere into the OS/cache/network/buffers.

The send function for POSIX sockets is not guaranteed to transmit all data. In C++ `boost::asio::async_write` adds the send operation to a work queue and returns immediately.

To support disconnection after ABORT, I needed to add special logic in my implementation that performs the disconnection in the `async_write` completion handler. Even then, I'm not sure of the operating system's behavior: can a socket disconnect operation cut off a network send still in progress? Should I add a delay between the end of `async_send` and socket disconnect?

**Comment by KSDaemon** [comment]
Okay I understand that. But how it is done now in your implementation for cases of incorrect message types (like getting a second WELCOME)?

**Comment by ecorm** [comment]
> Okay I understand that. But how it is done now in your implementation for cases of incorrect message types (like getting a second WELCOME)?

It currently sends an ABORT message and disconnects the transport socket when the networking library tells it the message has been completely sent (whatever that means at the socket/OS/network interface level). I haven't yet tested if the client fully receives the ABORT message before disconnection. Anyways. my C++ router implementation is still a work in progress, so I don't mind changing that protocol error handling behavior to match the WAMP standard.

I'm on the fence about whether ABORT messages should be sent as a result of protocol errors. I see both the advantages and disadvantages and can't decide for myself what's best. :thinking:  Should this be up to the router implementation to decide, or mandated by the standard?

**Comment by ecorm** [comment]
> Okay I understand that. But how it is done now in your implementation for cases of incorrect message types (like getting a second WELCOME)?

Oh, sorry, I see now that you're asking what I currently do in my _client_ implementation. I current disconnect abruptly without sending an ABORT message.

I'm overhauling the CppWAMP client implementation at the same time that I'm writing the new router implementation. I don't mind changing the client behavior if the standard mandates that ABORT messages be sent in response to protocol errors.

The user base for CppWAMP is very small, so I don't mind making changes to better follow the standard. I'm sure that most of the clones I see in the GitHub stats are just robots. :grin:

**Comment by ecorm** [comment]
> So even if you close the network connection from your high-level app code, the message will already be in network subsystem queues or even on the hardware level which can handle retransmission or some other possible problems.

I found the relevant SO question [here](https://stackoverflow.com/questions/8874021), and the recommendation in the linked [blog post](https://blog.netherlabs.nl/articles/2009/01/18/the-ultimate-so_linger-page-or-why-is-my-tcp-not-reliable) is to use the `shutdown()` system call after the message is completely sent. This sends a TCP FIN packet which causes the remote peer to close the connection after it has received the final message. The local peer then just waits for the socket to be closed by the remote host, or closes it manually after a time out.

The networking library I use [exposes `shutdown()` in its API](https://www.boost.org/doc/libs/release/doc/html/boost_asio/reference/basic_stream_socket/shutdown.html), so I can use the above procedure.

I already have to deal with this problem for `ABORT` messages sent during the `HELLO` establishment phase, so it's not any more trouble to do the same for protocol errors.

It's still not clear to me if the transport is supposed to be disconnected after an `ABORT` is sent during session establishment due to authentication failure. Should a router allow a client to attempt another WAMP session establishment via a new `HELLO`->`CHALLENGE`->`AUTHENTICATE`->`WELCOME` sequence using the same transport session?

**Comment by SunFulong** [comment]
Hi all, I'm back!

I had planned to ask for a new PR after this PR was merged.

In next PR, I want to add new sections for chapters 4~6, which is an example of a complete story for each chapter. That will cover cases where the flow chart is not clear.

Maybe now is the time to submit the next PR?

**Comment by oberstet** [comment]
> I already have to deal with this problem for ABORT messages sent during the HELLO establishment phase, so it's not any more trouble to do the same for protocol errors.

those are different cases. crossbar will

1. send ABORT in case of authentication failure during WAMP opening handshake
2. drop the connection upon protocol error without sending ABORT before

> It's still not clear to me if the transport is supposed to be disconnected after an `ABORT` is sent during session establishment due to authentication failure.

"sending ABORT" should be read as: the outgoing data has been put into the socket

whether that data will still hit the wire when the socket is immediately dropped (the connection closed) is not relevant, as the client must cope with TCP connections becoming unreactive anyways - however, _if_ the client is still getting data before the connection is gone, that data must be ABORT

> Even then, I'm not sure of the operating system's behavior: can a socket disconnect operation cut off a network send still in progress? 

I don't think Posix has that specified, and Linux .. not sure. There are 2 cases you can check with wireshark:

* RST vs
* FIN + ACK

https://stackoverflow.com/questions/4160347/close-vs-shutdown-socket

**Comment by SunFulong** [comment]
> BUT: the [description of ABORT](https://wamp-proto.org/wamp_bp_latest_ietf.html#name-abort-2) only has a message sequence chart that shows ABORT as being used as part of the opening of a WAMP session, and the whole ABORT message is only introduced as a subsection of "4.1 Session Establishment"

I committed a new PR #439 , added an example cover ABORT message

**Pull request merged**